driver-core/src/main/java/com/datastax/driver/core/Metadata.java

```diff
@@ -633,20 +660,20 @@ private TokenMap(Token.Factory factory,
                 Host host = entry.getKey();
                 host.setTokens(ImmutableSet.copyOf(entry.getValue()));
             }
+            this.primaryToTokens = primaryToTokens;
         }
 
-        public static TokenMap build(Token.Factory factory, Map<Host, Collection<String>> allTokens, Collection<KeyspaceMetadata> keyspaces) {
+        public static TokenMap build(Token.Factory factory, Map<Host, ? extends Collection<Token>> allTokens, Collection<KeyspaceMetadata> keyspaces) {
```

Comment:
I like the idea of a simplified computation when only the keyspaces have changed (not the node list). But there's still a lot of keyspace-independent stuff that's recomputed here, basically everything except `hostsToRanges` (even `replStrategyToHosts` would be nice to cache between invocations). I wonder if we should not cache more aggressively to avoid this.